src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs

```diff
@@ -156,15 +158,16 @@ public static void MatrixTimesSource(bool transpose, AlignedArray matrix, Aligne
         /// </summary>
         /// <param name="value">The value to add.</param>
         /// <param name="destination">The destination to add the value to.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static void Add(float value, Span<float> destination)
         {
             Contracts.AssertNonEmpty(destination);
 
-            if (Avx.IsSupported)
+            if (Avx.IsSupported && destination.Length >= MinVectorSize)
             {
                 AvxIntrinsics.AddScalarU(value, destination);
             }
-            else if (Sse.IsSupported)
+            else if (Sse.IsSupported && destination.Length >= MinVectorSize)
```

Comment:
I would actually expect this to be beneficial for `4` or more elements, possibly with a check to special-case this scenario in order to avoid the "leading" vs "trailing" element handling...

Reply:
> I would actually expect this to be beneficial for `4` or more elements,

The benchmarks were still slower with MinVectorSize = 4

> possibly with a check to special-case this scenario in order to avoid the "leading" vs "trailing" element handling.

i didnt understand this. can you explain it little bit more ? Did you mean to do go for the software implementation for trailing element ?

Reply:
Take `Scale` as an example. It currently has 5-6 checks for the average code path, which allows it to  deal with with leading elements, the main block, and trailing elements.

For large inputs, this is fine and the branch checks are amortized. However, for small inputs, these branches add up and could be considered unnecessary (in cases like 4 elements, the branches are also likely more than the cost of the code actually executing).

I would expect that 4 or more elements is still the cutoff, but due to the current code structure, the various leading/trailing/alignment checks cause the code to look worse. Perhaps if we special-cased inputs that are between 4-16 elements (or some profiled upper bound) and only dealt with it via unaligned loads for these small cases, it would still be faster (on average) than the scalar code path.

Reply:
> Perhaps if we special-cased inputs that are between 4-16 elements (or some profiled upper bound) and only dealt with it via unaligned loads for these small cases, it would still be faster (on average) than the scalar code path.

yes we can try doing that. But we are using this updated algorithm of trailing and leading elements, only for sum and scale. Before that we should first need to move other functions to this new implementation. For this PR, i suggest just adding this check